packages/grafana-ui/src/components/Logs/LogRowContextProvider.tsx

```diff
@@ -58,69 +60,74 @@ export const getRowContexts = async (
-            // logs which were logged in the same timestamp and that can be a problem depending on the precision.
-            if (parseInt(timestampField.values.get(fieldIndex), 10) === row.timeEpochMs) {
-              continue;
-            }
+    const data: any[] = [];
+    for (let index = 0; index < dataResult.data.length; index++) {
+      const dataFrame = toDataFrame(dataResult.data[index]);
+      const fieldCache = new FieldCache(dataFrame);
+      const timestampField: Field<string> = fieldCache.getFieldByName('ts')!;
+      const idField: Field<string> | undefined = fieldCache.getFieldByName('id');
+
+      for (let fieldIndex = 0; fieldIndex < timestampField.values.length; fieldIndex++) {
+        // TODO: this filtering is datasource dependant so it will make sense to move it there so the API is
+        //  to return correct list of lines handling inclusive ranges or how to filter the correct line on the
+        //  datasource.
+
+        // Filter out the row that is the one used as a focal point for the context as we will get it in one of the
+        // requests.
+        if (idField) {
+          // For Loki this means we filter only the one row. Issue is we could have other rows logged at the same
+          // ns which came before but they come in the response that search for logs after. This means right now
+          // we will show those as if they came after. This is not strictly correct but seems better than losing them
+          // and making this correct would mean quite a bit of complexity to shuffle things around and messing up
+          //counts.
+          if (idField.values.get(fieldIndex) === row.uid) {
+            continue;
+          }
+        } else {
+          // Fallback to timestamp. This should not happen right now as this feature is implemented only for loki
+          // and that has ID. Later this branch could be used in other DS but mind that this could also filter out
+          // logs which were logged in the same timestamp and that can be a problem depending on the precision.
+          if (parseInt(timestampField.values.get(fieldIndex), 10) === row.timeEpochMs) {
+            continue;
           }
+        }
 
-          const lineField: Field<string> = dataFrame.fields.filter(field => field.name === 'line')[0];
-          const line = lineField.values.get(fieldIndex); // assuming that both fields have same length
+        const lineField: Field<string> = dataFrame.fields.filter(field => field.name === 'line')[0];
+        const line = lineField.values.get(fieldIndex); // assuming that both fields have same length
 
-          data.push(line);
-        }
+        data.push(line);
       }
+    }
 
-      return data;
-    }),
-    errors: results.map(result => {
-      const errorResult: DataQueryError = result as DataQueryError;
-      if (!errorResult.message) {
-        return '';
-      }
+    return logsSortOrder === LogsSortOrder.Ascending ? data.reverse() : data;
```

Comment:
If sort order if `ascending` (flipped), we have to reverse the order of logs in both - `before` and `after` arrays.